src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringFailoverPluginTest.java

```diff
@@ -0,0 +1,312 @@
+import org.mockito.MockitoAnnotations;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Stream;
+
+class NodeMonitoringFailoverPluginTest {
+
+  @Mock
+  private NodeMonitoringFailoverPlugin.IMonitorServiceInitializer monitorServiceInitializer;
+  @Mock
+  private PropertySet propertySet;
+  @Mock
+  private HostInfo hostInfo;
+  @Mock
+  private IFailoverPlugin mockPlugin;
+  @Mock
+  private Log logger;
+  @Mock
+  private IMonitorService monitorService;
+  @Mock
+  private Callable sqlFunc;
+  @Mock
+  private MonitorConnectionContext context;
+
+  @Mock
+  private RuntimeProperty<Boolean> nativeFailureDetectionEnabledProperty;
+  @Mock
+  private RuntimeProperty<Integer> failureDetectionTimeProperty;
+  @Mock
+  private RuntimeProperty<Integer> failureDetectionIntervalProperty;
+  @Mock
+  private RuntimeProperty<Integer> failureDetectionCountProperty;
+
+
+  private static final String NODE = "node";
+  private static final String NO_MONITOR_METHOD_NAME = "foo";
+  private static final String MONITOR_METHOD_NAME = "executeQuery";
+
+  private static final int FAILURE_DETECTION_TIME = 10;
+  private static final int FAILURE_DETECTION_INTERVAL = 100;
+  private static final int FAILURE_DETECTION_COUNT = 5;
+  private static final ExecutionException EXECUTION_EXCEPTION = new ExecutionException(
+      "exception",
+      new InvocationException("exception", new Throwable()));
+
+  private NodeMonitoringFailoverPlugin plugin;
+  private AutoCloseable closeable;
+
+  @BeforeEach
+  void init() {
+    closeable = MockitoAnnotations.openMocks(this);
+    plugin = new NodeMonitoringFailoverPlugin();
+
+    initDefaultMockReturns();
+  }
+
+  @AfterEach
+  void cleanUp() throws Exception {
+    closeable.close();
+  }
+
+  @ParameterizedTest
+  @MethodSource("generateNullArguments")
+  void test_1_initWithNullArguments(
+      final PropertySet set,
+      final HostInfo info,
+      final IFailoverPlugin failoverPlugin,
+      final Log log) {
+    Assertions.assertThrows(
+        NullArgumentException.class,
+        () -> plugin.init(set, info, failoverPlugin, log));
+  }
+
+  @Test
+  void test_2_initWithFailoverEnabled() {
+    Mockito
+        .when(nativeFailureDetectionEnabledProperty.getValue())
+        .thenReturn(Boolean.TRUE);
+
+    initializePlugin();
+
+    Mockito.verify(monitorServiceInitializer).create(Mockito.eq(logger));
+  }
+
+  @Test
+  void test_3_initWithFailoverDisabled() {
+    Mockito
+        .when(nativeFailureDetectionEnabledProperty.getValue())
+        .thenReturn(Boolean.FALSE);
+
+    initializePlugin();
+    Mockito.verify(monitorServiceInitializer, Mockito.never()).create(Mockito.any());
+  }
+
+  @Test
+  void test_4_executeWithFailoverDisabled() {
+    Mockito
+        .when(nativeFailureDetectionEnabledProperty.getValue())
+        .thenReturn(Boolean.FALSE);
+    initializePlugin();
+
+    Assertions.assertDoesNotThrow(() -> {
```

Comment:
Do we need assertDoesNotThrow? If you execute the code, and it doesn't throw an exception, that should be OK no?

Reply:
I think it makes the goal of the test more obvious, but I am okay with either approach.